File: Platforms/Windows/Reign.Video.D3D9/VertexBuffer.cpp
#include "pch.h"
#include "VertexBuffer.h"

namespace Reign
{namespace Video
{namespace D3D9
{
	#pragma region Properties
	VertexBufferTopologys VertexBuffer::Topology::get() {return topology;}
	void VertexBuffer::Topology::set(VertexBufferTopologys value)
	{
		switch (value)
		{
			case (VertexBufferTopologys::None): primitiveTopology = D3DPT_POINTLIST; primitiveVertexCount = 0; break;
			case (VertexBufferTopologys::Point): primitiveTopology = D3DPT_POINTLIST; primitiveVertexCount = 1; break;
			case (VertexBufferTopologys::Line): primitiveTopology = D3DPT_LINELIST; primitiveVertexCount = 2; break;
			case (VertexBufferTopologys::Triangle): primitiveTopology = D3DPT_TRIANGLELIST; primitiveVertexCount = 3; break;
		}
		topology = value;
	}
	#pragma endregion

	#pragma region Constructors
	VertexBuffer::VertexBuffer(DisposableI^ parent, BufferLayoutDescI^ bufferLayoutDesc, BufferUsages bufferUsage, VertexBufferTopologys vertexBufferTopology)
	: VertexBufferI(parent, bufferLayoutDesc, bufferUsage)
	{
		init(parent, vertexBufferTopology, nullptr);
	}

	VertexBuffer::VertexBuffer(DisposableI^ parent, BufferLayoutDescI^ bufferLayoutDesc, BufferUsages bufferUsage, VertexBufferTopologys vertexBufferTopology, array<float>^ vertices)
	: VertexBufferI(parent, bufferLayoutDesc, bufferUsage)
	{
		init(parent, vertexBufferTopology, vertices);
	}

	void VertexBuffer::init(DisposableI^ parent, VertexBufferTopologys vertexBufferTopology, array<float>^ vertices)
	{
		null();
		try
		{
			video = parent->FindParentOrSelfWithException<Video^>();
			Topology = vertexBufferTopology;
			
			if (vertices) Init(vertices);
		}
		catch (Exception^ ex)
		{
			delete this;
			throw ex;
		}
	}

	VertexBuffer::~VertexBuffer()
	{
		disposeChilderen();
		if (vertexBuffer) vertexBuffer->Release();
		null();
	}

	void VertexBuffer::null()
	{
		vertexBuffer = 0;
		video = nullptr;
	}
	#pragma endregion

	#pragma region Methods
	void VertexBuffer::Init(array<float>^ vertices)
	{
		VertexBufferI::Init(vertices);
		if (vertexBuffer)
		{
			vertexBuffer->Release();
			vertexBuffer = 0;
		}

		uint bufferSize = vertexCount * vertexByteSize;
		IDirect3DVertexBuffer9* vertexBufferTEMP;
		if (FAILED(video->Device->CreateVertexBuffer(bufferSize, D3DUSAGE_WRITEONLY, 0, video->IsExDevice ? D3DPOOL_DEFAULT : D3DPOOL_MANAGED, &vertexBufferTEMP, 0)))
		{
			Debug::ThrowError(L"VertexBuffer", L"Failed to create VertexBuffer");
		}
		vertexBuffer = vertexBufferTEMP;
		Update(vertices, vertexCount);
	}

	void VertexBuffer::Update(array<float>^ vertices, int updateCount)
	{
		void *data = 0;
		uint bufferSize = vertexByteSize * updateCount;
		vertexBuffer->Lock(0, bufferSize, (void**)&data, 0);
		pin_ptr<float>verticesTEMP = &vertices[0];
		memcpy(data, verticesTEMP, bufferSize);
		vertexBuffer->Unlock();
	}

	void VertexBuffer::enable(IndexBufferI^ indexBuffer, VertexBufferI^ instanceBuffer)
	{
		video->Device->SetStreamSource(0, vertexBuffer, 0, vertexByteSize);

		if (instanceBuffer)
		{
			VertexBuffer^ ib = (VertexBuffer^)instanceBuffer;
			video->Device->SetStreamSourceFreq(1, D3DSTREAMSOURCE_INSTANCEDATA | 1);
			video->Device->SetStreamSource(1, ib->vertexBuffer, 0, ib->vertexByteSize);
		}
		else
		{
			video->Device->SetStreamSourceFreq(0, 1);
			video->Device->SetStreamSourceFreq(1, 1);
		}

		if (indexBuffer)
		{
			this->indexBuffer = (IndexBuffer^)indexBuffer;
			this->indexBuffer->enable();
		}
		else
		{
			this->indexBuffer = nullptr;
		}
	}

	void VertexBuffer::Enable()
	{
		enable(nullptr, nullptr);
	}

	void VertexBuffer::Enable(IndexBufferI^ indexBuffer)
	{
		enable(indexBuffer, nullptr);
	}

	void VertexBuffer::Enable(VertexBufferI^ instanceBuffer)
	{
		enable(nullptr, instanceBuffer);
	}

	void VertexBuffer::Enable(IndexBufferI^ indexBuffer, VertexBufferI^ instanceBuffer)
	{
		enable(indexBuffer, instanceBuffer);
	}

	void VertexBuffer::Draw()
	{
		if (indexBuffer == nullptr) video->Device->DrawPrimitive(primitiveTopology, 0, (vertexCount/primitiveVertexCount));
		else video->Device->DrawIndexedPrimitive(primitiveTopology, 0, 0, vertexCount, 0, (indexBuffer->IndexCount/primitiveVertexCount));
	}

	void VertexBuffer::Draw(int drawCount)
	{
		if (indexBuffer == nullptr) video->Device->DrawPrimitive(primitiveTopology, 0, (drawCount/primitiveVertexCount));
		else video->Device->DrawIndexedPrimitive(primitiveTopology, 0, 0, vertexCount, 0, (drawCount/primitiveVertexCount));
	}

	void VertexBuffer::DrawInstanced(int drawCount)
	{
		video->Device->SetStreamSourceFreq(0, D3DSTREAMSOURCE_INDEXEDDATA | drawCount);
		Draw();
	}

	void VertexBuffer::DrawInstancedClassic(int drawCount, int meshVertexCount, int meshIndexCount)
	{
		if (indexBuffer == nullptr) video->Device->DrawPrimitive(primitiveTopology, 0, drawCount * (meshVertexCount/primitiveVertexCount));
		else video->Device->DrawIndexedPrimitive(primitiveTopology, 0, 0, drawCount * meshVertexCount, 0, drawCount * (meshIndexCount/primitiveVertexCount));
	}
	#pragma endregion
}
}
}

File: Platforms/Common/Reign.Video.D3D11/Buffers/VertexBuffer.cs
using Reign.Core;
using Reign_Video_D3D11_Component;
using System;

namespace Reign.Video.D3D11
{
	public class VertexBuffer : VertexBufferI
	{
		#region Properties
		private VertexBufferCom com;

		private VertexBufferTopologys topology;
		public override VertexBufferTopologys Topology
		{
			get {return topology;}
			set
			{
				switch (value)
				{
					case VertexBufferTopologys.Triangle: com.SetTopology(REIGN_D3D_PRIMITIVE_TOPOLOGY.TRIANGLELIST); break;
					case VertexBufferTopologys.Line: com.SetTopology(REIGN_D3D_PRIMITIVE_TOPOLOGY.LINELIST); break;
					case VertexBufferTopologys.Point: com.SetTopology(REIGN_D3D_PRIMITIVE_TOPOLOGY.POINTLIST); break;
				}
				topology = value;
			}
		}

		private IndexBuffer indexBuffer;
		public override IndexBufferI IndexBuffer
		{
			get {return indexBuffer;}
		}
		#endregion

		#region Constructors
		public static VertexBuffer New(DisposableI parent, BufferLayoutDescI bufferLayoutDesc, BufferUsages usage, VertexBufferTopologys topology, float[] vertices)
		{
			return new VertexBuffer(parent, bufferLayoutDesc, usage, topology, vertices);
		}

		public static VertexBuffer New(DisposableI parent, BufferLayoutDescI bufferLayoutDesc, BufferUsages usage, VertexBufferTopologys topology, float[] vertices, int[] indices)
		{
			return new VertexBuffer(parent, bufferLayoutDesc, usage, topology, vertices, indices);
		}

		public VertexBuffer(DisposableI parent, BufferLayoutDescI bufferLayoutDesc, BufferUsages usage, VertexBufferTopologys topology, float[] vertices)
		: base(parent, bufferLayoutDesc, usage, vertices)
		{
			init(parent, bufferLayoutDesc, usage, topology, vertices, null);
		}

		public VertexBuffer(DisposableI parent, BufferLayoutDescI bufferLayoutDesc, BufferUsages usage, VertexBufferTopologys topology, float[] vertices, int[] indices)
		: base(parent, bufferLayoutDesc, usage, vertices)
		{
			init(parent, bufferLayoutDesc, usage, topology, vertices, indices);
		}

		private void init(DisposableI parent, BufferLayoutDescI bufferLayoutDesc, BufferUsages usage, VertexBufferTopologys topology, float[] vertices, int[] indices)
		{
			try
			{
				var video = parent.FindParentOrSelfWithException<Video>();

				this.topology = topology;
				REIGN_D3D_PRIMITIVE_TOPOLOGY topologyType = REIGN_D3D_PRIMITIVE_TOPOLOGY.TRIANGLELIST;
				switch (topology)
				{
					case VertexBufferTopologys.Triangle: topologyType = REIGN_D3D_PRIMITIVE_TOPOLOGY.TRIANGLELIST; break;
					case VertexBufferTopologys.Line: topologyType = REIGN_D3D_PRIMITIVE_TOPOLOGY.LINELIST; break;
					case VertexBufferTopologys.Point: topologyType = REIGN_D3D_PRIMITIVE_TOPOLOGY.POINTLIST; break;
				}

				com = new VertexBufferCom(video.com, topologyType);
				var bufferUsage = (usage == BufferUsages.Write) ? REIGN_D3D11_USAGE.DYNAMIC : REIGN_D3D11_USAGE.DEFAULT;
				var cpuUsage = (usage == BufferUsages.Write) ? REIGN_D3D11_CPU_ACCESS_FLAG.WRITE : (REIGN_D3D11_CPU_ACCESS_FLAG)0;
				var error = com.Init(vertices, vertexCount, vertexByteSize, bufferUsage, cpuUsage);
				if (error == VertexBufferErrors.Buffer) Debug.ThrowError("VertexBuffer", "Failed to create Buffer");

				if (indices != null && indices.Length != 0) indexBuffer = new IndexBuffer(this, usage, indices);
			}
			catch (Exception e)
			{
				Dispose();
				throw e;
			}
		}

		public override void Dispose()
		{
			disposeChilderen();
			if (com != null)
			{
				com.Dispose();
				com = null;
			}
			base.Dispose();
		}
		#endregion

		#region Methods
		public override void Update(float[] vertices, int updateCount)
		{
			com.Update(vertices, updateCount);
		}

		public override void Enable()
		{
			if (indexBuffer != null) com.Enable(indexBuffer.com, null);
			else com.Enable(null, null);
		}

		public override void Enable(IndexBufferI indexBuffer)
		{
			com.Enable(((IndexBuffer)indexBuffer).com, null);
		}

		public override void Enable(VertexBufferI instanceBuffer)
		{
			if (indexBuffer != null) com.Enable(indexBuffer.com, ((VertexBuffer)instanceBuffer).com);
			else com.Enable(null, ((VertexBuffer)instanceBuffer).com);
		}

		public override void Enable(IndexBufferI indexBuffer, VertexBufferI instanceBuffer)
		{
			com.Enable(((IndexBuffer)indexBuffer).com, ((VertexBuffer)instanceBuffer).com);
		}

		public override void Draw()
		{
			com.Draw();
		}

		public override void Draw(int drawCount)
		{
			com.Draw(drawCount);
		}

		public override void DrawInstanced(int drawCount)
		{
			com.DrawInstanced(drawCount);
		}

		public override void DrawInstancedClassic(int drawCount, int meshVertexCount, int meshIndexCount)
		{
			com.DrawInstancedClassic(drawCount, meshVertexCount, meshIndexCount);
		}
		#endregion
	}
}